File: app/src/main/java/com/czbix/v2ex/model/Comment.java
package com.czbix.v2ex.model;

import android.os.Parcel;
import android.os.Parcelable;

import com.czbix.v2ex.network.RequestHelper;
import com.google.common.base.Objects;

public class Comment implements Parcelable, Thankable, Ignorable {
    private final int mId;
    private final String mContent;
    private final Member mMember;
    private final String mReplyTime;
    private final int mThanks;
    private final int mFloor;
    private final boolean mThanked;

    Comment(int id, String content, Member member, String replyTime, int thanks, int floor, boolean thanked) {
        mId = id;
        mContent = content;
        mMember = member;
        mReplyTime = replyTime;
        mThanks = thanks;
        mFloor = floor;
        mThanked = thanked;
    }

    public int getId() {
        return mId;
    }

    public String getContent() {
        return mContent;
    }

    public Member getMember() {
        return mMember;
    }

    public String getReplyTime() {
        return mReplyTime;
    }

    public int getThanks() {
        return mThanks;
    }

    public int getFloor() {
        return mFloor;
    }

    public boolean isThanked() {
        return mThanked;
    }

    @Override
    public String getIgnoreUrl() {
        return String.format("%s/ignore/reply/%d", RequestHelper.BASE_URL, mId);
    }

    @Override
    public String getThankUrl() {
        return String.format("%s/thank/reply/%d", RequestHelper.BASE_URL, mId);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Comment)) return false;
        Comment comment = (Comment) o;
        return Objects.equal(mId, comment.mId) &&
                Objects.equal(mThanks, comment.mThanks) &&
                Objects.equal(mContent, comment.mContent) &&
                Objects.equal(mReplyTime, comment.mReplyTime);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(mId, mContent, mReplyTime, mThanks);
    }

    public static class Builder {
        private int mId;
        private String mContent;
        private Member mMember;
        private String mReplyTime;
        private int mThanks;
        private int mFloor;
        private boolean mThanked;

        public Builder setId(int id) {
            mId = id;
            return this;
        }

        public Builder setContent(String content) {
            mContent = content;
            return this;
        }

        public Builder setMember(Member member) {
            mMember = member;
            return this;
        }

        public Builder setReplyTime(String time) {
            mReplyTime = time;
            return this;
        }

        public Builder setThanks(int thanks) {
            mThanks = thanks;
            return this;
        }

        public Builder setFloor(int floor) {
            mFloor = floor;
            return this;
        }

        public Builder setThanked(boolean thanked) {
            mThanked = thanked;
            return this;
        }

        public Comment createComment() {
            return new Comment(mId, mContent, mMember, mReplyTime, mThanks, mFloor, mThanked);
        }
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(this.mId);
        dest.writeString(this.mContent);
        dest.writeParcelable(this.mMember, 0);
        dest.writeString(this.mReplyTime);
        dest.writeInt(this.mThanks);
        dest.writeInt(this.mFloor);
        dest.writeByte(mThanked ? (byte) 1 : (byte) 0);
    }

    protected Comment(Parcel in) {
        this.mId = in.readInt();
        this.mContent = in.readString();
        this.mMember = in.readParcelable(Member.class.getClassLoader());
        this.mReplyTime = in.readString();
        this.mThanks = in.readInt();
        this.mFloor = in.readInt();
        this.mThanked = in.readByte() != 0;
    }

    public static final Creator<Comment> CREATOR = new Creator<Comment>() {
        public Comment createFromParcel(Parcel source) {
            return new Comment(source);
        }

        public Comment[] newArray(int size) {
            return new Comment[size];
        }
    };
}

File: app/src/main/kotlin/com/czbix/v2ex/db/Comment.kt
package com.czbix.v2ex.db

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.czbix.v2ex.model.Ignorable
import com.czbix.v2ex.model.Thankable
import com.czbix.v2ex.network.RequestHelper

@Entity
data class Comment(
        @PrimaryKey
        val id: Int,
        @ColumnInfo(index = true)
        val topicId: Int,
        val page: Int,
        val content: String,
        val username: String,
        val addAt: String,
        val thanks: Int,
        val floor: Int,
        val thanked: Boolean
) : Thankable, Ignorable {

    override fun getIgnoreUrl(): String {
        return String.format("%s/ignore/reply/%d", RequestHelper.BASE_URL, id)
    }

    override fun getThankUrl(): String {
        return String.format("%s/thank/reply/%d", RequestHelper.BASE_URL, id)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Comment

        if (id != other.id) return false
        if (topicId != other.topicId) return false
        if (page != other.page) return false
        if (content != other.content) return false
        if (username != other.username) return false
        if (addAt != other.addAt) return false
        if (thanks != other.thanks) return false
        if (floor != other.floor) return false
        if (thanked != other.thanked) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id
        result = 31 * result + topicId
        result = 31 * result + page
        result = 31 * result + content.hashCode()
        result = 31 * result + username.hashCode()
        result = 31 * result + addAt.hashCode()
        result = 31 * result + thanks
        result = 31 * result + floor
        result = 31 * result + thanked.hashCode()
        return result
    }

    class Builder {
        var topicId: Int = 0
        var id: Int = 0
        var page: Int = 0
        lateinit var content: String
        lateinit var username: String
        lateinit var addAt: String
        var thanks: Int = 0
        var floor: Int = 0
        var thanked: Boolean = false

        fun build(): Comment {
            check(topicId > 0)
            check(id > 0)
            check(floor > 0)
            check(page > 0)

            return Comment(id, topicId, page, content, username, addAt, thanks, floor, thanked)
        }
    }
}
